### PR TITLE
VB-2693 - updates to send SMS when visit booked, updated or cancelled.

### PR DESCRIPTION
**Pull request by dhirajshettymoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/NotificationService.kt [hunk 1]
```diff
@@ -2,24 +2,129 @@ package uk.gov.justice.digital.hmpps.notificationsalertsvsip.service
...
+      }
+    }
+  }
+
+  private fun getVisit(bookingReference: String): VisitDto? {
+    return try {
+      visitSchedulerClient.getVisitByReference(bookingReference)
+    } catch (e: Exception) {
+      // if there was an error getting the visit return null
+      LOG.error("no visit found for visit with booking reference - $bookingReference")
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/NotificationService.kt:90]: what needs to happen if the visit is not found? is it ok if an SMS is not sent in this case?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/config/WebClientConfiguration.kt [hunk 2]
```diff
@@ -29,6 +37,18 @@ class WebClientConfiguration(
     return getWebClient(hmppsAuthUrl, oauth2Client)
   }
 
+  @Bean
+  fun visitSchedulerWebClient(authorizedClientManager: OAuth2AuthorizedClientManager): WebClient {
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/config/WebClientConfiguration.kt:41]: getVisitSchedulerWebClient or createVisitSchedulerWebClient

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/config/WebClientConfiguration.kt [hunk 3]
```diff
@@ -29,6 +37,18 @@ class WebClientConfiguration(
...
+    val oauth2Client = getOauth2Client(authorizedClientManager, HmppsAuthClientRegistrationId.VISIT_SCHEDULER.clientRegistrationId)
+    return getWebClient(visitSchedulerBaseUrl, oauth2Client)
+  }
+
+  @Bean
+  fun prisonRegisterWebClient(authorizedClientManager: OAuth2AuthorizedClientManager): WebClient {
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/config/WebClientConfiguration.kt:47]: getPrisonRegisterWebClient or createPrisonRegisterWebClient

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/NotificationService.kt (hunk 1 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/NotificationService.kt:90]: I think a log is the only thing you can do, if it's not there,.. may be tie it down to the type of exception 404 log and forget other issues throw exception on.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/integration/domainevents/PrisonVisitsEventsTest.kt [hunk 4]
```diff
@@ -1,53 +1,246 @@
 package uk.gov.justice.digital.hmpps.notificationsalertsvsip.integration.domainevents
 
+import org.assertj.core.api.Assertions
```
Comment (AledWEvans) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/integration/domainevents/PrisonVisitsEventsTest.kt:3]: a lot of duplication in your test, would try and consolidate into a couple of test methods

**Review by AledWEvans — COMMENTED** [review]
Nice work some comments

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/config/WebClientConfiguration.kt (hunk 2 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/config/WebClientConfiguration.kt:41]: that is also the bean name which is used by qualifier annotation in VisitSchedulerClient so I think visitSchedulerWebClient workd better and it is also consistent with what we have in the Orchestration service.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/config/WebClientConfiguration.kt (hunk 3 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/config/WebClientConfiguration.kt:47]: consistent with Orchestration Service.

**Review by AledWEvans — APPROVED** [review]
LGTM

File: src/test/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/integration/domainevents/PrisonVisitsEventsTest.kt (hunk 4 above)
Comment (dhirajshettymoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/integration/domainevents/PrisonVisitsEventsTest.kt:3]: updated

**Pull request merged**